lib/optimize/SplitChunksPlugin.js

```diff
@@ -162,192 +297,216 @@ const totalSize = sizes => {
+ */
+const checkModuleType = (test, module) => {
+	if (test === undefined) return true;
+	if (typeof test === "function") {
+		return test(module.type);
+	}
+	if (typeof test === "string") {
+		const type = module.type;
+		return test === type;
 	}
+	if (test instanceof RegExp) {
+		const type = module.type;
+		return test.test(type);
+	}
+	return false;
+};
+
+/**
+ * @typedef {Object} CacheGroupKey
+ * @property {string} key
+ */
+
+/**
+ * @param {OptimizationSplitChunksCacheGroup&CacheGroupKey} options the group options
+ * @returns {CacheGroupSource} the normalized cached group
+ */
+const createCacheGroupSource = options => {
+	return {
+		key: options.key,
+		priority: options.priority,
+		getName: normalizeName(options.name) || defaultGetName,
+		chunksFilter: normalizeChunksFilter(options.chunks),
+		enforce: options.enforce,
+		minSize: normalizeSizes(options.minSize),
+		maxSize: normalizeSizes(options.maxSize),
+		minChunks: options.minChunks,
+		maxAsyncRequests: options.maxAsyncRequests,
+		maxInitialRequests: options.maxInitialRequests,
+		filename: options.filename,
+		idHint: options.idHint,
+		automaticNameDelimiter: options.automaticNameDelimiter,
+		reuseExistingChunk: options.reuseExistingChunk
+	};
+};
+
+module.exports = class SplitChunksPlugin {
 	/**
-	 * @param {undefined|string|RegExp|function(string): boolean} test type option
-	 * @param {Module} module the module
-	 * @returns {boolean} true, if the module should be selected
+	 * @param {OptimizationSplitChunksOptions=} options plugin options
 	 */
-	static checkModuleType(test, module) {
-		if (test === undefined) return true;
-		if (typeof test === "function") {
-			return test(module.type);
-		}
-		if (typeof test === "string") {
-			const type = module.type;
-			return test === type;
-		}
-		if (test instanceof RegExp) {
-			const type = module.type;
-			return test.test(type);
-		}
-		return false;
+	constructor(options = {}) {
+		const fallbackCacheGroup = options.fallbackCacheGroup || {};
+
+		/** @type {SplitChunksOptions} */
+		this.options = {
+			chunksFilter: normalizeChunksFilter(options.chunks || "all"),
+			minSize: normalizeSizes(options.minSize),
+			maxSize: normalizeSizes(options.maxSize),
+			maxInitialSize: normalizeSizes(options.maxInitialSize || options.maxSize),
+			maxAsyncSize: normalizeSizes(options.maxAsyncSize || options.maxSize),
+			minChunks: options.minChunks || 1,
+			maxAsyncRequests: options.maxAsyncRequests || 1,
+			maxInitialRequests: options.maxInitialRequests || 1,
+			hidePathInfo: options.hidePathInfo || false,
+			filename: options.filename || undefined,
+			getCacheGroups: normalizeCacheGroups(options.cacheGroups, options.name),
+			getName: normalizeName(options.name),
+			automaticNameDelimiter: options.automaticNameDelimiter,
+			fallbackCacheGroup: {
+				minSize: mergeSizes(fallbackCacheGroup.minSize, options.minSize),
+				maxSize: mergeSizes(fallbackCacheGroup.maxSize, options.maxSize),
```

Comment:
Normalization to `maxAsyncSize` and `maxInitialSize`

``` js
maxAsyncSize: mergeSizes(
  fallbackCacheGroup.maxAsyncSize,
  fallbackCacheGroup.maxSize,
  options.maxAsyncSize,
  options.maxSize
)
```

This is similar to normal cache groups, but normal cache groups do it in two steps.